database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java

```diff
@@ -302,40 +310,60 @@ private Schema inferSchema(Class<? extends Driver> driverClass) {
                                           dbSinkConfig.getJdbcPluginName());
       Properties connectionProperties = new Properties();
       connectionProperties.putAll(dbSinkConfig.getConnectionArguments());
-      try (Connection connection = DriverManager.getConnection(dbSinkConfig.getConnectionString(),
-                                                               connectionProperties)) {
+      try (Connection connection = createConnectionWithRetry(connectionProperties)) {
         executeInitQueries(connection, dbSinkConfig.getInitQueries());
-
-        try (Statement statement = connection.createStatement();
-             ResultSet rs = statement.executeQuery("SELECT * FROM " + fullyQualifiedTableName
-                                                     + " WHERE 1 = 0")) {
+        try (Statement statement = createStatementWithRetry(connection);
+             ResultSet rs = Failsafe.with(retryPolicy).<ResultSet>get(() -> statement
+               .executeQuery("SELECT * FROM " + fullyQualifiedTableName + " WHERE 1 = 0"))) {
           inferredFields.addAll(getSchemaReader().getSchemaFields(rs));
         }
-      } catch (SQLException e) {
-        // wrap exception to ensure SQLException-child instances not exposed to contexts w/o jdbc driver in classpath
-        String errorMessage =
-          String.format("SQL Exception occurred: [Message='%s', SQLState='%s', ErrorCode='%s'].", e.getMessage(),
-            e.getSQLState(), e.getErrorCode());
-        String errorMessageWithDetails = String.format("Error while reading table metadata." +
-          "Error message: '%s'. Error code: '%s'. SQLState: '%s'", e.getMessage(), e.getErrorCode(), e.getSQLState());
-        String externalDocumentationLink = getExternalDocumentationLink();
-        if (!Strings.isNullOrEmpty(externalDocumentationLink)) {
-          if (!errorMessage.endsWith(".")) {
-            errorMessage = errorMessage + ".";
-          }
-          errorMessage = String.format("%s For more details, see %s", errorMessageWithDetails, errorMessage);
+      } catch (FailsafeException e) {
+        Throwable cause = e.getCause();
+        if (cause instanceof SQLException) {
+          throw wrapException((SQLException) cause);
         }
-        throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
-          errorMessage, errorMessageWithDetails, ErrorType.USER, false, ErrorCodeType.SQLSTATE,
-          e.getSQLState(), externalDocumentationLink, new SQLException(e.getMessage(),
-            e.getSQLState(), e.getErrorCode()));
+        throw e;
```

Comment:
we should throw `cause` instead of `FailsafeException`.

Reply:
Handled the `FailsafeException` in a separate method : `unwrapFailsafeException` of `RetryUtils` class as suggested in the other comment.